File: backend/app/modules/Auth/auth_service.py
import uuid
import logging
import secrets
from datetime import datetime, timedelta

from sqlalchemy.ext.asyncio import AsyncSession

from app.utils.results import Result
from app.modules.Auth.auth_repository import AuthRepository
from app.modules.Auth.auth_schema import UserCreate, LoginRequest
from app.db.models.user import User
from app.db.models.types import UserRole
from app.utils.security import (
    hash_password,
    verify_password,
    create_access_token,
    create_refresh_token,
)
from app.config.settings import settings  # Only one settings import
from app.utils.email_service import EmailService

logger = logging.getLogger(__name__)

# Store reset tokens in memory
reset_tokens: dict = {}

class AuthService:
    def __init__(self, db: AsyncSession):
        self.db = db
        self.auth_repo = AuthRepository(db)
    
    async def register_nanny(self, user: UserCreate) -> Result:
        try:
            if await self.auth_repo.get_user_by_email(user.email):
                return Result.fail("Email already exists.", status_code=400)
            
            hashed_pass = hash_password(user.password)
            user_data = user.model_dump()
            user_data["password"] = hashed_pass
            user_data["role"] = UserRole.NANNY
            
            new_user = User(**user_data)
            self.db.add(new_user)
            await self.db.commit()
            await self.db.refresh(new_user)
            
            return Result.ok(data=new_user, status_code=201)
        except Exception as e:
            await self.db.rollback()
            print(f"DEBUG ERROR: {str(e)}") 
            return Result.fail(f"Internal Database Error: {str(e)}", status_code=500)

    async def register_family(self, user: UserCreate) -> Result:
        try:
            existing_user = await self.auth_repo.get_user_by_email(user.email)
            if existing_user:
                return Result.fail("User with this email already exists.", status_code=400)
            
            existing_phone = await self.auth_repo.get_user_by_phone(user.phone)
            if existing_phone:
                return Result.fail("User with this phone number already exists.", status_code=400)
            
            hashed_pass = hash_password(user.password)
            user.password = hashed_pass
            user.role = UserRole.FAMILY
            
            new_user = await self.auth_repo.create_user(user)
            await self.db.commit()
            await self.db.refresh(new_user)
            
            return Result.ok(data=new_user, status_code=201)
        except Exception as e:
            await self.db.rollback()
            return Result.fail(f"Registration failed: {str(e)}", status_code=500)
    
    async def login_user(self, form_data: LoginRequest) -> Result:
        user = await self.auth_repo.get_user_by_email(form_data.email)
        if not user:
            return Result.fail("User with this email does not exist", status_code=401)
        
        if not verify_password(form_data.password, user.password):
            return Result.fail("Invalid credentials", status_code=400)

        token_payload = {
            "sub": str(user.id), 
            "role": user.role.value if hasattr(user.role, 'value') else str(user.role)
        }

        access_token = create_access_token(
            data=token_payload, 
            expires_delta=timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
        )
        
        refresh_token = create_refresh_token(
            data=token_payload, 
            expires_delta=timedelta(days=settings.REFRESH_TOKEN_EXPIRE_DAYS)
        )

        user_data = {
            "id": str(user.id),
            "email": user.email,
            "role": user.role,
            "access_token": access_token,
            "refresh_token": refresh_token, 
            "token_type": "bearer"
        }
        return Result.ok(data=user_data, status_code=200)
    
    async def request_password_reset(self, email: str) -> Result:
        """Generate a password reset token and send email"""
        try:
            user = await self.auth_repo.get_user_by_email(email)
            if not user:
                return Result.ok(data={"message": "If your email is registered, you will receive a reset link."})
            
            token = secrets.token_urlsafe(32)
            
            reset_tokens[token] = {
                "user_id": str(user.id),
                "expires_at": datetime.utcnow() + timedelta(minutes=30)
            }
            
            # Print to console for debugging
            reset_link = f"{settings.FRONTEND_URL}/views/reset-password.html?token={token}"
            print("\n" + "="*60)
            print(f"PASSWORD RESET LINK FOR {email}:")
            print(reset_link)
            print("="*60 + "\n")
            
            # Try to send email
            try:
                email_sent = await EmailService.send_password_reset_email(email, token)
                if email_sent:
                    return Result.ok(data={"message": "Password reset link has been sent to your email."})
                else:
                    return Result.ok(data={"message": "Check the console for the reset link (development mode)."})
            except Exception as e:
                logger.warning(f"Email sending failed: {e}")
                return Result.ok(data={"message": "Check the console for the reset link (development mode)."})
                
        except Exception as e:
            logger.error(f"Password reset request error: {str(e)}")
            return Result.fail(f"Failed to process reset request", status_code=500)
        
    async def confirm_password_reset(self, token: str, new_password: str) -> Result:
        """Validate token and update password"""
        try:
            token_data = reset_tokens.get(token)
            if not token_data:
                return Result.fail("Invalid or expired reset token", status_code=400)
            
            if datetime.utcnow() > token_data["expires_at"]:
                del reset_tokens[token]
                return Result.fail("Reset token has expired. Please request a new one.", status_code=400)
            
            hashed_password = hash_password(new_password)
            user_id = uuid.UUID(token_data["user_id"])
            success = await self.auth_repo.update_user_password(user_id, hashed_password)
            
            if success:
                del reset_tokens[token]
                await self.db.commit()
                return Result.ok(data={"message": "Password has been reset successfully"})
            else:
                return Result.fail("User not found", status_code=404)
                
        except Exception as e:
            await self.db.rollback()
            return Result.fail(f"Failed to reset password: {str(e)}", status_code=500)

File: frontend/src/utils/authService.js
import { API_URL } from "../utils/config.js";

/**
 * Log in an existing user.
 * Endpoint: POST /login
 * Body: { email, password }
 * @param {string} emailAddress
 * @param {string} password
 * @returns {{ success: boolean, access_token?: string, message?: string }}
 */
export async function login(emailAddress, password) {
  try {
    const response = await fetch(`${API_URL}/login`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ email: emailAddress, password }),
    });

    const responseData = await response.json();

    if (response.ok) {
      return { success: true, ...responseData };
    } else {
      return {
        success: false,
        message: responseData.detail || "Login failed. Check your credentials.",
      };
    }
  } catch (networkError) {
    console.error("[authService] login network error:", networkError);
    return { success: false, message: "Network error. Please check your connection." };
  }
}

/**
 * Register a new user.
 * Endpoint: POST /nanny  (role === "nanny")
 *           POST /family (role === "family")
 * @param {{ email, password, first_name, last_name, role }} userData
 * @returns {{ success: boolean, data?: object, message?: string }}
 */
export async function signup(userData) {
  try {
    const roleEndpoint = userData.role?.toLowerCase() === "nanny" ? "nanny" : "family";

    const response = await fetch(`${API_URL}/${roleEndpoint}`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(userData),
    });

    const responseData = await response.json();

    if (response.ok) {
      return { success: true, data: responseData };
    } else {
      return {
        success: false,
        message: responseData.detail || "Signup failed. Please try again.",
      };
    }
  } catch (networkError) {
    console.error("[authService] signup network error:", networkError);
    return { success: false, message: "Network error occurred. Please try again." };
  }
}

/**
 * Request a password reset email.
 * Endpoint: POST /auth/forgot-password
 * @param {string} emailAddress
 * @returns {{ success: boolean, message: string }}
 */
export async function requestPasswordReset(emailAddress) {
  try {
    const response = await fetch(`${API_URL}/auth/forgot-password`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ email: emailAddress }),
    });

    const responseData = await response.json();

    if (response.ok) {
      return {
        success: true,
        message: responseData.message || "Reset link sent. Check your email.",
      };
    } else {
      return {
        success: false,
        message: responseData.detail || "Could not send reset email.",
      };
    }
  } catch (networkError) {
    console.error("[authService] requestPasswordReset network error:", networkError);
    return { success: false, message: "Network error. Please try again." };
  }
}

/**
 * Confirm a password reset with the token from the email link.
 * Endpoint: POST /auth/reset-password
 * @param {string} resetToken  — from URL param ?token=...
 * @param {string} newPassword
 * @returns {{ success: boolean, message: string }}
 */
export async function confirmPasswordReset(resetToken, newPassword) {
  try {
    const response = await fetch(`${API_URL}/auth/reset-password`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ token: resetToken, new_password: newPassword }),
    });

    const responseData = await response.json();

    if (response.ok) {
      return {
        success: true,
        message: responseData.message || "Password reset successfully.",
      };
    } else {
      return {
        success: false,
        message: responseData.detail || "Reset failed. The link may have expired.",
      };
    }
  } catch (networkError) {
    console.error("[authService] confirmPasswordReset network error:", networkError);
    return { success: false, message: "Network error. Please try again." };
  }
}